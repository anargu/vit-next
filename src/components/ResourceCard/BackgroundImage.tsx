
export type BackgroundImageProps = {
  src : string,
  alt : string,
};

export const BackgroundImage = (props : BackgroundImageProps) => (
  <div
    role="img"
    alt={props.alt}
    className={`w-full h-full bg-contain bg-no-repeat`}
    style={{ backgroundImage: `url(${props.src})`}}
  />
);
