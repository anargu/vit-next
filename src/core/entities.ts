import { ONE_DAY_IN_MILLIS, ONE_HOUR_IN_MILLIS, ONE_MINUTE_IN_MILLIS, ONE_MONTH_IN_MILLIS } from "./constants";

export interface VITResource {
  id?: string;
  og_image : string | null,
  keyphrase : string | null,
  date_created : string,
  og_title : string | null,
  og_description : string | null,
  url_title: string | null,
  url: string | null,
};

export class Resource {
  id : string;
  imageSrc : string | null;
  imageAlt : string | null;
  postedAt : Date;
  title : string | null;
  description : string | null;

  constructor(id : string, { og_image, keyphrase, date_created, og_title, url_title, og_description, url } : VITResource) {
    this.id = id;
    this.imageSrc = og_image;
    this.imageAlt = keyphrase;
    // Input date_created is in ISO 8601 date format/
    this.postedAt = new Date(date_created);
    this.title = og_title ?? url_title ?? url;
    this.description = og_description;
  }

  get timeAgo() : SinceDatetimeField {
    return new SinceDatetimeField(this.postedAt);
  };

  static fromVITResource (data : VITResource) : Resource {
    return new Resource("", data);
  };
};

export class SinceDatetimeField {

  value : string;

  constructor(date : Date | string | null) {
    if (typeof date === "string" || date === null) {
      this.value = "";
      return;
    } 

    const now = new Date();

    const diffTimeMillis = now.getTime() - date?.getTime();

    this.value = this.computeTime(diffTimeMillis);
  }

  computeTime (diffTimeMillis : number) : string {

    if (diffTimeMillis < ONE_MINUTE_IN_MILLIS)
      return "now";

    if (diffTimeMillis < ONE_HOUR_IN_MILLIS)
      return `${Math.floor(diffTimeMillis / ONE_MINUTE_IN_MILLIS)} m ago`;

    if (diffTimeMillis < ONE_DAY_IN_MILLIS)
      return `${Math.floor(diffTimeMillis / ONE_HOUR_IN_MILLIS)} h ago`;

    if (diffTimeMillis < ONE_MONTH_IN_MILLIS)
      return `${Math.floor(diffTimeMillis / ONE_DAY_IN_MILLIS)} days ago`;

    return `${Math.floor(diffTimeMillis / ONE_MONTH_IN_MILLIS)} months ago`;
  }

  toString () : string { return this.value; }
}

export interface User {
  id : string;
  email : string;
  displayName : string;
}

export type AuthenticatedUser = User;

export interface Link {
  isDeleted : boolean;
}



