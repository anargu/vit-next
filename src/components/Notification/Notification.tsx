import React from 'react';
import { NotificationsProvider, showNotification as mantineNotification } from '@mantine/notifications';
import { DefaultMantineColor } from '@mantine/core';

export type NotificationProps = {
  children?: JSX.Element;
};

export type NotificationOptions = {
  color?: DefaultMantineColor | undefined | string,
}

export const showNotification = (message: string, title : string = "Information", options: NotificationOptions) => {
  mantineNotification({
    title: title,
    message: message,
    color: options?.color ?? "indigo",
  });
};

export const showDefaultNotification = (message: string, title : string = "Information") => {
  mantineNotification({
    message,
    title,
    color: "indigo",
  });
};


export const WithNotificationsProvider = (props : NotificationProps) : JSX.Element => {
  if (!props.children) return <div />;

  return <NotificationsProvider>{props.children}</NotificationsProvider>;
};
