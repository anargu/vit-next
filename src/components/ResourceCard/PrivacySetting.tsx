
import { Switch } from '@mantine/core';
import React from 'react'

export type PrivacySettingProps = {
  isPublic : boolean,
  isDisabled? : boolean,
  onChangePrivacy: (newValue : boolean) => void | Promise<void>,
};

export const PrivacySetting = (props : PrivacySettingProps) => {
  return (
    <div>
      <Switch
        label="Make it public"
        checked={props.isPublic}
        disabled={props.isDisabled}
        onChange={(event) => props.onChangePrivacy(event.currentTarget.checked)}
      />
    </div>
  )
};

