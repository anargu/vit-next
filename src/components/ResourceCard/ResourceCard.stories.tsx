import React from 'react';
import { ComponentStory, ComponentMeta } from '@storybook/react';

import { ResourceCard, ResourceCardProps } from "./ResourceCard";

import { faker } from "@faker-js/faker";
import { VITResource } from '@/src/core/entities';
// More on default export: https://storybook.js.org/docs/react/writing-stories/introduction#default-export
export default {
  title: 'UI/ResourceCard',
  component: ResourceCard,
  // More on argTypes: https://storybook.js.org/docs/react/api/argtypes
  argTypes: {
  },
} as ComponentMeta<typeof ResourceCard>;

// More on component templates: https://storybook.js.org/docs/react/writing-stories/introduction#using-args
const Template: ComponentStory<typeof ResourceCard> = (args : ResourceCardProps) => (
  <div className="bg-gray-100 w-[300px] h-[400px]"><ResourceCard {...args} /></div>
);

const mockHitData = () : VITResource => ({
  id: faker.datatype.uuid(),
  og_image: "https://source.unsplash.com/random/50x50",
  keyphrase: "man holding a beer",
  date_created: faker.datatype.datetime().toISOString(),
  og_title: "Inter Link",
  og_description: "This is a link, interlinked",
  url_title: "The link",
  url: "https://www.google.com",
});

export const Default = Template.bind({});
// More on args: https://storybook.js.org/docs/react/writing-stories/args
Default.args = {
  hit: mockHitData(),
};

export const LongText = Template.bind({});

LongText.args = {
  hit: {
    ...mockHitData(),
    og_title: faker.lorem.paragraph(2),
    og_description: faker.lorem.paragraphs(4),
  }
};

const FluidTemplate: ComponentStory<typeof ResourceCard> = (args : ResourceCardProps) => (
  <div className="bg-gray-100 h-[400px]"><ResourceCard {...args} /></div>
);

export const FluidLayout = FluidTemplate.bind({});

FluidLayout.args = {
  hit: {
    ...mockHitData(),
    og_title: faker.lorem.paragraph(2),
    og_description: faker.lorem.paragraphs(4),
  }
};

export const NoImage = Template.bind({});
// More on args: https://storybook.js.org/docs/react/writing-stories/args
NoImage.args = {
  hit: {
    ...mockHitData(),
    og_image: null,
  }
};
