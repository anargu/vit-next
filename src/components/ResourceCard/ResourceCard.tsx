import { Resource, VITResource } from "../../core/entities";
import Image from "next/image"
import { useMemo } from "react";
import styled from 'styled-components';
import Link from "../../../public/assets/link.svg";
import Bookmark from "../../../public/assets/bookmark.svg";
import Share from "../../../public/assets/share.svg";

export type ResourceCardProps = {
  hit: VITResource,
};

export const ResourceCard = ({ hit } : ResourceCardProps) => {

  const resourceData = useMemo(() => Resource.fromVITResource(hit), [hit]);

  return (
    <div className="relative py-4 px-4 text-white h-full">
      <div className="absolute top-0 right-0 bottom-0 left-0 bg-black/60 z-20">
      </div>
      <div className="absolute top-0 right-0 bottom-0 left-0 z-10">
        <Image src={resourceData.imageSrc} alt={resourceData.imageAlt} layout="fill"/>
      </div>

      <div className="relative z-30 h-full grid grid-rows-[min-content_2fr_1fr_36px] gap-5">
        <div className="text-right"><span>{resourceData.timeAgo.toString()}</span></div>

        <p className={`py-0 max-h-[calc(2.5rem*4)]
          text-4xl text-multiline-overflow-mixin
          text-ellipsis [-webkit-line-clamp:4]`}>{resourceData.title}</p>

        <p className="self-center max-h-[calc(2rem*4)] text-2xl text-multiline-overflow-mixin text-ellipsis [-webkit-line-clamp:2]">{resourceData.description}</p>
        <ActionsStyled className="flex justify-around">
          <Link/>
          <Bookmark/>
          <Share/>
        </ActionsStyled>
      </div>
    </div>
  );
};

const ActionsStyled = styled.div`
  svg path {
    stroke: #FFF;
  }
`;
