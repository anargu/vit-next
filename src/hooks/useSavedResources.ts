import { useEffect, useState } from "react";
import { SAVED_LINK_KEY } from "../components/ResourceCard/ResourceCard";
import { VITResource } from "../core/entities";

export const useSavedResources = () => {

  const [savedResources, setSavedResources] = useState<null | VITResource[]>(null);

  useEffect(() => {
    if (!!savedResources) return;

    const rawData = localStorage.getItem(SAVED_LINK_KEY);
    if (!rawData) return setSavedResources([]);
    
    const savedPosts : VITResource[] | null = JSON.parse(rawData);
    if (!savedPosts) return setSavedResources([]);

    /* const _resources = savedPosts.map((post) => Resource.fromVITResource(post)); */
    
    setSavedResources(savedPosts);
  }, [savedResources]);

  return { savedResources };
};
