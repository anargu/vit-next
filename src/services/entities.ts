import { Resource } from "../core/entities";
import { Timestamp } from "../firebase";

export type Link = {
  id : string | null;
  deleted: boolean;
  isPublic: boolean | undefined;

  createdAt: Timestamp;
  updatedAt: Timestamp;
  postedAt : Timestamp;

  url: string;
  imageSrc : string | null;
  imageAlt : string | null;
  title : string | null;
  description : string | null;
};

export class DataTransformer {
  static fromLinktoResource(srcData : Link) : Resource {

    const resource = new Resource({
      id: srcData.id ?? undefined,
      date_created : srcData.createdAt.toDate().toUTCString(),
      date_updated : srcData.updatedAt.toDate().toUTCString(),
      deleted: srcData.deleted,
      is_public: srcData.isPublic ?? false,

      og_image : srcData.imageSrc,
      keyphrase : srcData.imageAlt,
      og_title : srcData.title,
      og_description : srcData.description,
      url_title: srcData.title,
      url: srcData.url,
    });

    return resource;
  }
}
