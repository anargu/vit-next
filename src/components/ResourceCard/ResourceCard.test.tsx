import { VITResource } from "@/src/core/entities";
import { useSavedResources } from "@/src/hooks/useSavedResources";
import { faker } from "@faker-js/faker";
import { MantineProvider } from "@mantine/core";
import { userEvent } from "@storybook/testing-library";
import { act, findByRole, render } from "@testing-library/react";
import { WithNotificationsProvider } from "../Notification/Notification";
import { ResourceCard, SAVED_LINK_KEY } from "./ResourceCard";

faker.seed(1);

Object.assign(navigator, {
  clipboard: {
    writeText: () => {},
  },
});

export const patternPostedAt = /\d+\s(m|h|hours?|days?|months?|years?)\sago/;

describe("ResourceCard", () => {

  const mockedResource = () : VITResource => ({
    id: faker.datatype.uuid(),
    og_image: "https://source.unsplash.com/random/50x50",
    keyphrase: faker.lorem.sentence(5),
    date_created: faker.datatype.datetime().toISOString(),
    og_title: faker.lorem.sentence(4),
    og_description: faker.lorem.sentence(12),
    url: faker.internet.url(),
    url_title: faker.lorem.sentence(4),
  });

  it("renders the component", async  () => {
    const { container } = render(<ResourceCard hit={mockedResource()} />);
    expect(container).toMatchSnapshot();
  });

  it("should appear DetailedCard Fullscreen Dialog when user taps on card", async  () => {
    const resourceData = { ...mockedResource() };
    const { findByText, container } = render(<ResourceCard hit={resourceData} />);

    const resourceCard = container.firstElementChild;
    userEvent.click(resourceCard!);

    const element = await findByText("Visit Site");
    expect(element).not.toBeNull();
  });

  describe("image field", () => {
    it("displays an image", async  () => {
      const resourceData = { ...mockedResource() };
      const { findByRole } = render(<ResourceCard hit={resourceData} />);

      const imageElement = findByRole("img");
      expect(imageElement).not.toBeFalsy();
    });

    it("renders no image if there is not src url", async  () => {
      const resourceData = { ...mockedResource(), og_image: null, keyphrase: null };
      const { container } = render(<ResourceCard hit={resourceData} />);

      const results = container.querySelectorAll("img");

      expect(results.length).toBe(0);
    });
  });

  describe("Title field", () => {

    it("displays title", async () => {

      const resourceData = mockedResource();

      const { findByText } = render(<ResourceCard hit={resourceData} />);

      const titleElement = await findByText(resourceData.og_title!);

      expect(titleElement).not.toBeFalsy();
    });

    it("displays url_title if title is null", async () => {
      const URL_TITLE = "this is a URL title";
      const hit : VITResource = { ...mockedResource(), og_title: null, url_title: URL_TITLE };

      const { findByText } = render(<ResourceCard hit={hit} />);

      const titleElement = await findByText(URL_TITLE);

      expect(titleElement).not.toBeFalsy();
    });

    it("displays url if title and url_title are not set", async () => {
      const URL = "https://www.google.com";
      const hit : VITResource = { ...mockedResource(), og_title: null, url_title: null, url: URL };

      const { findByText } = render(<ResourceCard hit={hit} />);

      const titleElement = await findByText(URL);

      expect(titleElement).not.toBeFalsy();
    });
  });

  describe("field postedAt", () => {

    it("displays \"now\" when postedAt diff with current time is lower than a minute", async () => {
      const hit = mockedResource();
      hit.date_created = new Date().toISOString();

      const { findByText } = render(<ResourceCard hit={hit} />);

      const postedAtElement = await findByText("now");

      expect(postedAtElement).not.toBeFalsy();
    });

    it("displays created time diff with current time if diff is greater than a minute ", async () => {
      const hit = mockedResource();
      const yesterday = new Date();
      yesterday.setDate(yesterday.getDate() - 1);

      hit.date_created = yesterday.toISOString();
      const { findByText } = render(<ResourceCard hit={hit} />);

      const postedAtElement = await findByText(patternPostedAt);

      expect(postedAtElement).not.toBeFalsy();
    });
  });

  describe("Actions", () => {

    // it("display a more detailed page when clicked on card but not ", async () => {
    //   const { findByText } = render(<ResourceCard hit={mockedResource()} />);
    //
    //   expect(await findByText("Visit site")).not.toBeFalsy();
    // });

    describe("Save", () => {

      it("post is saved on local storage among others previous saved posts when save button is clicked", async () => {
        const oldPost = mockedResource();
        localStorage.setItem(SAVED_LINK_KEY, JSON.stringify([oldPost]));

        const newPost = mockedResource();

        const WrapperResourceCard = () => {
          const { saveResource } = useSavedResources();

          return (
            <ResourceCard onSaveResource={saveResource} hit={newPost} />
          )
        };

        const { findByTitle, findByText } = render(
          <MantineProvider>
            <WithNotificationsProvider>
              <WrapperResourceCard/>
            </WithNotificationsProvider>
          </MantineProvider>
        );

        await act(async () => {
          const saveButtonEl = await findByTitle("Save Button");
          saveButtonEl.click();
        });

        await findByText("Link saved locally.");
        const value = localStorage.getItem(SAVED_LINK_KEY);
        expect(value).not.toBeFalsy();
        expect(value).not.toBeNull();

        const actualSavedPosts = JSON.parse(value!)
        expect(actualSavedPosts).toHaveLength(2);
        expect(actualSavedPosts[0]).toStrictEqual(oldPost);
        expect(actualSavedPosts[1]).toStrictEqual(newPost);
      });

      it("post is saved on empty local storage when save button is clicked ", async () => {
        localStorage.removeItem(SAVED_LINK_KEY);

        const WrapperResourceCard = () => {
          const { saveResource } = useSavedResources();

          return (
            <ResourceCard onSaveResource={saveResource} hit={mockedResource()} />
          )
        };

        const { findByTitle, findByText } = render(
          <MantineProvider>
            <WithNotificationsProvider>
              <WrapperResourceCard/>
            </WithNotificationsProvider>
          </MantineProvider>
        );

        await act(async () => {
          const saveButtonEl = await findByTitle("Save Button");
          saveButtonEl.click();
        });

        // Check if Notification is shown.
        await findByText("Link saved locally.");
        const value = localStorage.getItem(SAVED_LINK_KEY);
        expect(value).not.toBeFalsy();
      });

      afterEach(() => {
        localStorage.removeItem(SAVED_LINK_KEY);
      });
    });

    describe("Share", () => {
      jest.spyOn(navigator.clipboard, "writeText");

      it("link is copied on clipboard when share button is clicked ", async () => {
        const hit = mockedResource();
        hit.date_created = new Date().toISOString();

        const { findByTitle } = render(<ResourceCard hit={hit} />);

        await act(async () => {
          const shareButtonEl = await findByTitle("Share Button");
          shareButtonEl.click();
        });

        expect(navigator.clipboard.writeText).toHaveBeenCalledWith(hit.url);
      });
    });

  });
});
