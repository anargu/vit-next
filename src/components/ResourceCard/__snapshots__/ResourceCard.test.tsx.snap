// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResourceCard renders the component 1`] = `
<div>
  <div
    class=" rounded relative py-4 px-4 text-white h-full"
  >
    <div
      class=" rounded absolute top-0 right-0 bottom-0 left-0 bg-slate-900/80 z-20"
    />
    <div
      class="rounded overflow-hidden absolute top-0 right-0 bottom-0 left-0 z-10"
    >
      <div
        class="w-full h-full bg-cover bg-no-repeat"
        role="img"
        style="background-image: url(https://source.unsplash.com/random/50x50);"
      />
    </div>
    <div
      class="relative z-30 h-full grid grid-rows-[min-content_2fr_1fr_36px] gap-5"
    >
      <div
        class="text-right"
      >
        <span>
          216 months ago
        </span>
      </div>
      <p
        class="py-0 max-h-[calc(2.5rem*4)]
            text-xl text-multiline-overflow-mixin
            text-ellipsis [-webkit-line-clamp:4]"
      >
        Itaque velit omnis et.
      </p>
      <p
        class="self-center max-h-[calc(2rem*4)] text-lg text-multiline-overflow-mixin text-ellipsis [-webkit-line-clamp:2]"
      >
        Porro ut et ipsam explicabo eligendi occaecati debitis et saepe eum dicta.
      </p>
      <div
        class="sc-bczRLJ kQZhXl flex justify-around"
      >
        <span
          class="relative z-20"
          title="Save Button"
        >
          <svg />
        </span>
        <span
          title="Share Button"
        >
          <svg />
        </span>
      </div>
    </div>
  </div>
</div>
`;
